src/Tokenizers/PHP.php

```diff
@@ -588,6 +588,58 @@ protected function tokenize($string)
                 echo PHP_EOL;
             }
 
+            if ($tokenIsArray === true
+                && isset(Util\Tokens::$contextSensitiveKeywords[$token[0]]) === true
+                && isset($this->tstringContexts[$finalTokens[$lastNotEmptyToken]['code']]) === true
+            ) {
+                $preserveKeyword = false;
+
+                // `new class` should be preserved
+                if ($token[0] === T_CLASS && $finalTokens[$lastNotEmptyToken]['code'] === T_NEW) {
+                    $preserveKeyword = true;
+                }
+
+                // `new class implements` should be preserved
+                if ($finalTokens[$lastNotEmptyToken]['code'] === T_CLASS) {
+                    $preserveKeyword = true;
+                }
+
+                // `namespace\` should be preserved
```

Comment:
There's a test: https://github.com/squizlabs/PHP_CodeSniffer/pull/3484/commits/6f991b5a24d8ee36773a043c24f41156a0ed13ba#diff-884ebc6b1d508f7200a3ef0edc6dd662e97072d9fec37dd6a4e374a2d835f6d4R3-R4

Reply:
That's not what I meant (please unresolve this).

Think:
```php
namespace foreach;
namespace my\class\extensions;
```

Reply:
Added test case.

Reply:
Noted. That test case still only safeguards the first of the above examples though, the current tokenizer would still fail on the second.

Reply:
Added one more test :)

Reply:
Thanks. Would be wonderful though if the other parts of the namespace name would also be checked. The test currently only checks the first part, i.e. `my`.

What I'm trying to point out, is that on PHP < 8.0, this snippet:
```php
<?php

namespace my\class\extensions;
namespace my\foreach\yield;
```

... would be tokenized as - take note of `class`, `yield` and `foreach`:
```
Ptr | Ln | Col  | Cond | ( #) | Token Type                 | [len]: Content
-------------------------------------------------------------------------
  0 | L1 | C  1 | CC 0 | ( 0) | T_OPEN_TAG                 | [5]: <?php

  1 | L2 | C  1 | CC 0 | ( 0) | T_WHITESPACE               | [0]:

  2 | L3 | C  1 | CC 0 | ( 0) | T_NAMESPACE                | [9]: namespace
  3 | L3 | C 10 | CC 0 | ( 0) | T_WHITESPACE               | [1]:
  4 | L3 | C 11 | CC 0 | ( 0) | T_STRING                   | [2]: my
  5 | L3 | C 13 | CC 0 | ( 0) | T_NS_SEPARATOR             | [1]: \
  6 | L3 | C 14 | CC 0 | ( 0) | T_CLASS                    | [5]: class
  7 | L3 | C 19 | CC 0 | ( 0) | T_NS_SEPARATOR             | [1]: \
  8 | L3 | C 20 | CC 0 | ( 0) | T_STRING                   | [10]: extensions
  9 | L3 | C 30 | CC 0 | ( 0) | T_SEMICOLON                | [1]: ;
 10 | L3 | C 31 | CC 0 | ( 0) | T_WHITESPACE               | [0]:

 11 | L4 | C  1 | CC 0 | ( 0) | T_NAMESPACE                | [9]: namespace
 12 | L4 | C 10 | CC 0 | ( 0) | T_WHITESPACE               | [1]:
 13 | L4 | C 11 | CC 0 | ( 0) | T_STRING                   | [2]: my
 14 | L4 | C 13 | CC 0 | ( 0) | T_NS_SEPARATOR             | [1]: \
 15 | L4 | C 14 | CC 0 | ( 0) | T_FOREACH                  | [7]: foreach
 16 | L4 | C 21 | CC 0 | ( 0) | T_NS_SEPARATOR             | [1]: \
 17 | L4 | C 22 | CC 0 | ( 0) | T_YIELD                    | [5]: yield
 18 | L4 | C 27 | CC 0 | ( 0) | T_SEMICOLON                | [1]: ;
 19 | L4 | C 28 | CC 0 | ( 0) | T_WHITESPACE               | [0]:
```

... while with your change, it will be tokenized (better), like so:
```
Ptr | Ln | Col  | Cond | ( #) | Token Type                 | [len]: Content
-------------------------------------------------------------------------
  0 | L1 | C  1 | CC 0 | ( 0) | T_OPEN_TAG                 | [5]: <?php

  1 | L2 | C  1 | CC 0 | ( 0) | T_WHITESPACE               | [0]:

  2 | L3 | C  1 | CC 0 | ( 0) | T_NAMESPACE                | [9]: namespace
  3 | L3 | C 10 | CC 0 | ( 0) | T_WHITESPACE               | [1]:
  4 | L3 | C 11 | CC 0 | ( 0) | T_STRING                   | [2]: my
  5 | L3 | C 13 | CC 0 | ( 0) | T_NS_SEPARATOR             | [1]: \
  6 | L3 | C 14 | CC 0 | ( 0) | T_STRING                   | [5]: class
  7 | L3 | C 19 | CC 0 | ( 0) | T_NS_SEPARATOR             | [1]: \
  8 | L3 | C 20 | CC 0 | ( 0) | T_STRING                   | [10]: extensions
  9 | L3 | C 30 | CC 0 | ( 0) | T_SEMICOLON                | [1]: ;
 10 | L3 | C 31 | CC 0 | ( 0) | T_WHITESPACE               | [0]:

 11 | L4 | C  1 | CC 0 | ( 0) | T_NAMESPACE                | [9]: namespace
 12 | L4 | C 10 | CC 0 | ( 0) | T_WHITESPACE               | [1]:
 13 | L4 | C 11 | CC 0 | ( 0) | T_STRING                   | [2]: my
 14 | L4 | C 13 | CC 0 | ( 0) | T_NS_SEPARATOR             | [1]: \
 15 | L4 | C 14 | CC 0 | ( 0) | T_STRING                   | [7]: foreach
 16 | L4 | C 21 | CC 0 | ( 0) | T_NS_SEPARATOR             | [1]: \
 17 | L4 | C 22 | CC 0 | ( 0) | T_STRING                   | [5]: yield
 18 | L4 | C 27 | CC 0 | ( 0) | T_SEMICOLON                | [1]: ;
 19 | L4 | C 28 | CC 0 | ( 0) | T_WHITESPACE               | [0]:
```

There is no difference on PHP 8.0+ as the "retokenization of namespaced names" already takes care of it in that case.

Reply:
Ok, now I finally get it :) Added.